core/unit_test/TestViewCtorProp.hpp

```diff
@@ -0,0 +1,145 @@
+  static_assert(std::is_default_constructible_v<vcp_label_base_t>);
+
+  static_assert(std::is_constructible_v<vcp_label_base_t, std::string>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, const std::string&>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, std::string&&>);
+
+  constexpr unsigned N = 3;
+  static_assert(std::is_constructible_v<vcp_label_base_t, const char[N]>);
+  static_assert(std::is_constructible_v<vcp_label_base_t, char[N]>);
+
+  static_assert(std::is_constructible_v<vcp_label_base_t, char*>);
+}
+
+// Check traits of derived class Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_traits) {
+  using vcp_label_base_t = Kokkos::Impl::ViewCtorProp<void, std::string>;
+  using vcp_label_t      = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  static_assert(std::is_base_of_v<vcp_label_base_t, vcp_label_t>);
+
+  static_assert(vcp_label_t::has_label);
+
+  // Check that the derived class is not default constructible. It is a design
+  // choice to not allow the default constructor to be called.
+  static_assert(!std::is_default_constructible_v<vcp_label_t>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, std::string>);
+  static_assert(std::is_constructible_v<vcp_label_t, const std::string&>);
+  static_assert(std::is_constructible_v<vcp_label_t, std::string&&>);
+
+  constexpr unsigned N = 3;
+  static_assert(std::is_constructible_v<vcp_label_t, const char[N]>);
+  static_assert(std::is_constructible_v<vcp_label_t, char[N]>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, char*>);
+}
+
+// Check that the constructor of Kokkos::Impl::ViewCtorProp<std::string>
+// moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_constructor_can_move) {
+  using vcp_label_t = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  std::string label("our label");
+
+  vcp_label_t prop(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check that Kokkos::view_alloc moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_view_alloc_can_move) {
+  std::string label("our label");
+
+  auto prop = Kokkos::view_alloc(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check the copy constructor of Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_copy_constructor) {
+  using vcp_empty_t = Kokkos::Impl::ViewCtorProp<>;
+  using vcp_label_t = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  // Copy construction from an empty view constructor properties object.
+  static_assert(std::is_constructible_v<vcp_label_t, const vcp_empty_t&>);
+
+  vcp_empty_t prop_empty;
+  vcp_label_t prop_empty_copy(prop_empty);
+
+  ASSERT_TRUE(
+      Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop_empty_copy)
+          .empty());
```

Comment:
Not sure what is the goal of these checks.
Just like default constructing a ViewCtorProp that contains a label, constructing from the base class makes little sense to me.

Reply:
The goal of this test is to ensure that use case like this works:
- https://github.com/kokkos/kokkos/blob/fa09fa8713d44cef31dcbb7bd5aace19a28a1244/core/src/View/Kokkos_ViewCtor.hpp#L267

Here, a view constructor properties object "with more properties" is copy constructed from a view constructor properties object "with less properties". The base classes holding the missing properties are default constructed.